pkg/cleanup/directories.go

```diff
@@ -43,30 +44,57 @@ func (d *directories) Run() error {
 		return err
 	}
 
+	var dataDirMounted bool
+
 	// search and unmount kubelet volume mounts
 	for _, v := range procMounts {
-		if strings.Compare(v.Path, fmt.Sprintf("%s/kubelet", d.Config.dataDir)) == 0 {
-			logrus.Debugf("%v is mounted! attempting to unmount...", v.Path)
-			if err = mounter.Unmount(v.Path); err != nil {
-				logrus.Warningf("failed to unmount %v", v.Path)
-			}
-		} else if strings.Compare(v.Path, d.Config.dataDir) == 0 {
+		if v.Path == filepath.Join(d.Config.dataDir, "kubelet") {
 			logrus.Debugf("%v is mounted! attempting to unmount...", v.Path)
 			if err = mounter.Unmount(v.Path); err != nil {
 				logrus.Warningf("failed to unmount %v", v.Path)
 			}
+		} else if v.Path == d.Config.dataDir {
+			dataDirMounted = true
 		}
 	}
 
-	logrus.Debugf("deleting k0s generated data-dir (%v) and run-dir (%v)", d.Config.dataDir, d.Config.runDir)
-	if err := os.RemoveAll(d.Config.dataDir); err != nil {
-		fmtError := fmt.Errorf("failed to delete %v. err: %v", d.Config.dataDir, err)
-		return fmtError
+	if dataDirMounted {
+		logrus.Debugf("removing contents of mounted data-dir (%v)", d.Config.dataDir)
+		if err := removeAllContents(d.Config.dataDir); err != nil {
+			return fmt.Errorf("failed to remove contents of %v. err: %v", d.Config.dataDir, err)
+		}
+	} else {
+		logrus.Debugf("deleting k0s generated data-dir (%v)", d.Config.dataDir)
+		if err := os.RemoveAll(d.Config.dataDir); err != nil {
+			return fmt.Errorf("failed to delete %v. err: %v", d.Config.dataDir, err)
+		}
 	}
```

Comment:
Had a second thought about this. There's another option that might save us from implementing directory traversal ourselves. We could simply rely on `os.RemoveAll`, as before, and inspect the error returned from it.

Something along those lines:

```go
logrus.Debugf("deleting k0s generated data-dir (%v)", d.Config.dataDir)
if err := os.RemoveAll(d.Config.dataDir); err != nil {
	if !dataDirMounted || !isMountDeletionErr(err, d.Config.dataDir) {
		return fmt.Errorf("failed to delete k0s generated data-dir %q: %w", d.Config.dataDir, err)
	}
}

...

func isMountDeletionErr(err error, dir string) bool {
	var pathErr *os.PathError
	return errors.As(err, &pathErr) &&
		pathErr.Path == dir &&
		pathErr.Op == "unlinkat" &&
		(pathErr.Err == syscall.EPERM || pathErr.Err == syscall.EACCES)
}
```